refactor(compiler-cli): setup compilation mode to enable generating linker code

JoostK opened the pull request:
This is a precursor to introducing the Angular linker. As an initial
step, a compiler option to configure the compilation model is introduced.
This option is initially internal until the linker is considered ready.

JoostK committed:
refactor(compiler-cli): setup compilation mode to enable generating linker code

This is a precursor to introducing the Angular linker. As an initial
step, a compiler option to configure the compilation mode is introduced.
This option is initially internal until the linker is considered ready.

JoostK commented on packages/compiler-cli/src/ngtsc/core/api/src/options.ts:
```diff
@@ -48,6 +48,22 @@ export interface TestOnlyOptions {
+   *
+   * @internal
+   */
+  compilationModel?: 'application'|'library';
```
Open to suggestions regarding naming here.

I choose `'application'` and `'library'` here instead of `'aot'` and `'prelink'` in the spike as this feels more intuitive from a user's perspective. One of the downsides of this naming scheme is that some libraries may want to use full AOT compilation even for libraries, e.g. during development. I can imagine there's another option to enable that mode, though.

Alternatively, we could go with a boolean option such as `'isLibrary'`, which could even be a tristate for the scenario I mentioned above:

- `isLibrary: false` (default)
- `isLibrary: true`
- `isLibrary: 'aot'`

As I said, open to suggestions :-)

petebacondarwin replied:
I feel like we should separate the compilation model from the final target. How about `full` and `partial`?

petebacondarwin replied:
Also, I feel like `CompilationMode` is more appropriate than `CompilationModel`?
But I would even go further and prefer something like `AotMode: Full|Partial`, especially since this has no effect in JIT mode.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/core/api/src/options.ts:
```diff
@@ -48,6 +48,22 @@ export interface TestOnlyOptions {
+ */
+export interface TargetOptions {
+  /**
+   * Specifies the compilation model to use. The following modes are available:
```
in fact you even use the term `mode` here :-)

petebacondarwin commented on packages/compiler-cli/src/ngtsc/transform/src/api.ts:
```diff
@@ -151,6 +166,16 @@ export interface DecoratorHandler<D, A, R> {
       constantPool: ConstantPool): CompileResult|CompileResult[];
+
+  /**
+   * Generates code for the decorator using a stable format suitable to be published to NPM.
```
Might be worth mentioning, here, that this generated output will be consumed by the linker.

petebacondarwin reviewed (commented):
Looking good - always room for discussion of naming :-)

I think it would be possible to make a test of the prelinking logic in the trait compiler in the `packages/compiler-cli/src/ngtsc/transform/test/compilation_spec.ts` file by creating a fake decorator handler that does actually implement the `compilerPrelink()` method.

JoostK committed:
test(compiler-cli): run compliance tests for two compilation modes

To verify the correctness of the linker output, we leverage the existing
compliance tests. The plan is to test the linker by running all compliance
tests using a full round trip of pre-linking and subsequently post-linking,
where the generated code should be identical to a full AOT compile.
To ensure that the output of the pre-linker is correct and stable, the
intermediate pre-linked generated code will be captured in golden files.

This commit introduces the ability to run the tests using two compilation
modes. Follow-up work is required to implement the logic for running
the linker round trip.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/transform/test/compilation_spec.ts:
```diff
@@ -43,12 +46,125 @@ runInEachFileSystem(() => {
+        expect(noPrelinkResult[0].name).toBe('compile');
+      });
+
+      it('should run regular compilation when implemented if compilation mode is library', () => {
```
```suggestion
      it('should run regular compilation when implemented if compilation mode is application', () => {
```

petebacondarwin commented on packages/compiler-cli/test/compliance/test_runner.ts:
```diff
@@ -0,0 +1,39 @@
+  });
+}
+
+export const runAllCompilationModels = runAllCompilationModesFn as RunAllCompilationModesFn;
```
How about

```suggestion
export const runForEachCompilationMode = runForEachCompilationModeFn as RunAllCompilationModesFn;
```

(similar to the FileSystem `runInEachFileSystem()`):

For example:
```ts
runForEachCompilationMode(compilationModel => {
  ...
});
```

gkalpak replied:
Why not `runInEachCompilationMode()`? :grin:

JoostK replied:
I did go with `runInEachCompilationMode`, just for consistency with the filesystem equivalent.

petebacondarwin reviewed (approved):
Looks good apart from the bike-shedding - a couple of minor nits.

petebacondarwin commented:
Oh and I would mark this as `target: minor` since `major` indicates a breaking change, right?

JoostK commented:
> Oh and I would mark this as `target: minor` since `major` indicates a breaking change, right?

You wrote the doc so you're probably the expert here. I targeted major as that's what master currently corresponds with, but I see how major becomes inaccurate if we were to transition master to be targeting 11.1.

petebacondarwin commented:
> > Oh and I would mark this as `target: minor` since `major` indicates a breaking change, right?
> 
> You wrote the doc so you're probably the expert here. I targeted major as that's what master currently corresponds with, but I see how major becomes inaccurate if we were to transition master to be targeting 11.1.

Hopefully it won't still be waiting to be merged when we get to 11.1.0 but just in case :-)

gkalpak reviewed (approved):
LGTM (sans the naming bike-shedding) :+1:

JoostK commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -479,8 +480,16 @@ export class TraitCompiler implements ProgramTypeCheckAdapter {
-      const compileMatchRes =
-          trait.handler.compile(clazz, trait.analysis, trait.resolution, constantPool);
+
+      let compileRes: CompileResult|CompileResult[];
```
I had to change `compileMatchRes` into `let` to allow separate assignment in the two branches, which resulted in TS no longer narrowing `compileMatchRes` to just `CompileResult` inside the `some` callback in the below `else-guard`. Having the intermediate variable allows for retaining the `const`-ness of `compileMatchRes` such that TS continues to narrow it as desired.

gkalpak replied:
That is really strange. Even with the explicit type:

```suggestion
      let compileMatchRes: CompileResult|CompileResult[];
```

JoostK replied:
I don't think it's that strange, although it does show a limitation of TypeScript's CFA. The explicit type declares the type to be either `CompilerResult` or `CompilerResult[]`, and the `if (Array.isArray(compileMatchRes))` branch below allows narrowing of `compileMatchRes` in the corresponding `else` guard. However, if `compileMatchRes` is no longer `const` then this narrowing is deemed unsafe inside the callback of `res.some(result => result.name === compileMatchRes.name)` as TS assumes that the closure may be called at a later point in time where the `Array.isArray` narrowing may no longer apply.

JoostK replied:
@petebacondarwin I think I tried both; that ternary doesn't really look nice to me and pulling it out into a separate method also felt cumbersome, due to the three parameters it would require. Hence I went with the intermediate variable. If anybody feels strongly about doing it differently then I'll update it :-) But I'm sticking with it for now.

petebacondarwin replied:
I'll not be holding out over this :-) (I do have a slight aversion to `let` in general though)

JoostK commented on packages/compiler-cli/src/ngtsc/transform/test/compilation_spec.ts:
```diff
@@ -43,12 +46,125 @@ runInEachFileSystem(() => {
+        expect(noPrelinkResult[0].name).toBe('compile');
+      });
+
+      it('should run regular compilation when implemented if compilation mode is library', () => {
```
Good point, will update.

petebacondarwin reviewed (approved):
LGTM! Quick turnaround @JoostK

petebacondarwin commented:
CI is not happy though...

JoostK commented:
I do have a PR that switches over to using two distinct Bazel targets: #38929. If we can get that ready soonish (probably needs to be done vastly differently to the state it is in now) we may want to go with that in instead of this one.

petebacondarwin commented:
Superceded by https://github.com/angular/angular/pull/38938
